docs/design/specs/runtime-async.md

```diff
@@ -0,0 +1,71 @@
+
+All async methods effectively have two entry points, or signatures. The first signature is the one present in the above code: a modopt before the return type. The second signature is a "Task-equivalent signature", described in further detail in [I.8.6.1.5 Method signatures].
+
+Async methods have a special calling convention and may not be called directly outside of other async methods. To call an async method from a sync method, callers must use the second "Task-equivalent signature".
+
+Callers may retrieve a Task-equivalent return type from an async method via calling the "Task-equivalent signature". This functionality is available in both sync and async methods.
+
+### I.8.6.1.5 Method signatures
+
+The list of relevant components is augmented to include sync vs. async method types. Async methods have some additions to normal signature compatibility.
+
+Async MethodDef entries implicitly create two member definitions: one explicit, primary definiton, and a second implicit, runtime-generated definition.
+
+The primary, mandatory definition must be present in metadata as a MethodDef. This signature must have a required custom modifier as the last custom modifier before the return type. The custom modifier must fit the following requirements:
+* If the return type is void, the custom modifier must be either `System.Threading.Task` or `System.Threading.ValueTask`.
+* If the return type is not `void`, the modifier must be to either `System.Threading.Task<T>` or `System.Threading.ValueTask<T>`. The return type must be a valid substitution for the type parameter of the custom modifier type.
+
+_[Note: async methods have the same return type conventions as sync methods. If the async method produces a System.Int32, the return type must be System.Int32.]_
+
+The second async signature is implicit and runtime-generated and is hereafter referred to as the "Task-equivalent" signature. It is generated based on the primary signature. The transformation is as follows:
+
+If the async return type is void, the return type of the Task-equivalent signature is the type of the async custom modifier.
+
+Otherwise, the Task-equivalent return type is the custom modifier type (either Task`1 or ValueTask`1), substituted with the async return type.
+
+### I.8.10.2 Method inheritance
+
+For the purposes of inheritance and hiding, both async signatures ([### I.8.6.1.5 Method signatures]) are used for hiding and overriding, but cannot be configured separately.
+
+### II.10.3.2 The .override directive
+
+Async methods participate in overrides through both definitions (both signatures). An async method with a .override overrides the target method signature, as well as the secondary "Task-equivalent" signature if applicable. An async method may also override only the "Task-equivalent" signature, if an async signature is not present on the base class. A sync method may not override an async method, even if it matches the async method's "Task-equivalent" definition.
```

Comment:
> A sync method may not override an async method, even if it matches the async method's "Task-equivalent" definition.

1.  What is a sync method in IL? (as opposed to an async1 method) How would you detect one?
2. I think this rule is saying that a nuget author can never start using async2 methods in their library until they're certain that every single downstream user has switched to async2.  Is this what we want?

Reply:
> A sync method may not override an async method, even if it matches the async method's "Task-equivalent" definition.

I think that is allowed. At least that is desirable.

Basically in most cases you can replace an old Task-returning method with a runtime-async implementation whose Task-based signature matches what it replaces. 
You are replacing one entry point with two, one of which is exactly the same as before.

Take-back changes when a runtime async implementation is replaced with an ordinary Task-returning method could be binary-breaking though. You are replacing two entry points with one and something may already be using another entry. 
In many cases even take-back changes can be fixed via a recompile (i.e. not source-breaking).

Reply:
I put it in that we will also allow overriding async methods with sync methods. I'm not completely sure what to do about interface implementations. The problem is that sometimes they're implicit. For async methods it's easy to say that every declaration in metadata represents two definitions in the type system, but sync methods are harder. I think it would be worrying to imply that every existing Task method is going to get a corresponding shadow async declaration. But for interface methods we would have to somehow recognize that a given Task method is the Task-equivalent signature for an interface, and that it can implement a method on an interface.